File: unmaintained/matlab/pi-exchange/pisender.m
% PISENDER script demonstrating MATLAB-HELICS interface
%
% Usage:
%  1. Start two separate MATLAB terminals 
%  2. In the first: 
%     >> pisender
%  3. In the second:
%     >> pireciever


%% Initialize HELICS library in MATLAB
helicsStartup()

import helics.*;
%% Configuration
deltat = 1;  %Base time interval (seconds)
numsteps = 20;

% HELICS options
% Note: these configure this matlab process to host the main broker and 1
% federate
pisend_start_broker = true;
helics_core_type = 'zmq'; 
broker_initstring = '--federates 2 --name=mainbroker';
fed_initstring = '--broker=mainbroker --federates=1';

%% Provide summary information

fprintf('PI SENDER (with main broker): Helics version = %s\n', helicsGetVersion())

%% Create broker (if desired)
if pisend_start_broker
    disp('Creating Broker');
    broker = helicsCreateBroker(helics_core_type, '', broker_initstring);
    disp('Created Broker');

    fprintf('Checking if Broker is connected...');
    isconnected = helics.helicsBrokerIsConnected(broker);

    if isconnected == 1
        fprintf('SUCCESS, Broker created and connected\n');
    else
        fprintf('\n')
        error('NOT CONNECTED (helicsBrokerIsConnected return = %d)', isconnected)
    end
end

%% Create Federate Info object that describes the federate properties
fedinfo = helicsCreateFederateInfo();
assert(not(isempty(fedinfo)))

% Set core type from string
helicsFederateInfoSetCoreTypeFromString(fedinfo, helics_core_type);


% Federate init string
helicsFederateInfoSetCoreInitString(fedinfo, fed_initstring);

% Note:
% HELICS minimum message time interval is 1 ns and by default
% it uses a time delta of 1 second. What is provided to the
% setTimedelta routine is a multiplier for the default timedelta 
% (default unit = seconds).

% Set one message interval
helicsFederateInfoSetTimeProperty(fedinfo,helics_property_time_delta,deltat);
helicsFederateInfoSetIntegerProperty(fedinfo,helics_property_int_log_level,helics_log_level_warning);

%% Actually create value federate
vfed = helicsCreateValueFederate('MATLAB Pi SENDER Federate',fedinfo);
disp('PI SENDER: Value federate created');

%% Register our value to publish
pub = helicsFederateRegisterGlobalPublication(vfed, 'testA', helics_data_type_double, '');
disp('PI SENDER: Publication registered (testA)');

%% Start execution
try
    helicsFederateEnterExecutingMode(vfed);
    disp('PI SENDER: Entering execution mode');
catch e
    error('PI SENDER: Failed to enter execution mode (status = X)\n Try running pisender.m first. (or start the broker seperately)');
end

%% Execution Loop
% This federate will be publishing deltat*pi for numsteps steps
this_time = 0.0;
value = pi;

for i = 1:numsteps
    val = value;

     granted_time = helicsFederateRequestTime(vfed, i);

%    fprintf('PI SENDER: Publishing value pi = %g at time %f\n', val, this_time + (deltat * i));
    fprintf('PI SENDER: Publishing value pi = %g at time %4.1f... ', val, granted_time);
  helicsPublicationPublishDouble(pub, val);
    fprintf('DONE\n');
end

%% Shutdown

if pisend_start_broker
    % If we started the broker in this thread, we have to be careful
    % sequencing the shutdown in hopes of doing so cleanly
    helicsFederateFinalize(vfed);
    disp('PI SENDER: Federate finalized');

    %Make sure the broker is gone in case we have a lingering low-level
    %reference (to avoid memory leaks)
    helicsBrokerWaitForDisconnect(broker,-1);
    disp('PI SENDER: Broker disconnected');

    helics.helicsFederateFree(vfed);
    helics.helicsCloseLibrary();
else
    %But if we just setup the federate, we can simply call endFederate
    helicsFederateDestroy(vfed); %#ok<UNRCH>  
    disp('PI SENDER: Federate finalized');
end

File: unmaintained/python/delayed-pi-exchange/pisender.py
# -*- coding: utf-8 -*-
import time
import helics as h
from math import pi

initstring = "-f 2 --name=mainbroker"
fedinitstring = "--broker=mainbroker --federates=1"
deltat = 0.01

helicsversion = h.helicsGetVersion()

print("PI SENDER: Helics version = {}".format(helicsversion))

# Create broker #
print("Creating Broker")
broker = h.helicsCreateBroker("zmq", "", initstring)
print("Created Broker")

print("Checking if Broker is connected")
isconnected = h.helicsBrokerIsConnected(broker)
print("Checked if Broker is connected")

if isconnected == 1:
    print("Broker created and connected")

# Create Federate Info object that describes the federate properties #
fedinfo = h.helicsCreateFederateInfo()

# Set Federate name #
h.helicsFederateInfoSetCoreName(fedinfo, "TestA Federate")

# Set core type from string #
h.helicsFederateInfoSetCoreTypeFromString(fedinfo, "zmq")

# Federate init string #
h.helicsFederateInfoSetCoreInitString(fedinfo, fedinitstring)

# Set the message interval (timedelta) for federate. Note th#
# HELICS minimum message time interval is 1 ns and by default
# it uses a time delta of 1 second. What is provided to the
# setTimedelta routine is a multiplier for the default timedelta.

# Set one second message interval #
h.helicsFederateInfoSetTimeProperty(fedinfo, h.helics_property_time_delta, deltat)

# Create value federate #
cfed = h.helicsCreateCombinationFederate("TestA Federate", fedinfo)
print("PI SENDER: Combination federate created")

# Register the publication #
pub = h.helicsFederateRegisterGlobalTypePublication(cfed, "testA", "double", "")
print("PI SENDER: Publication registered")


epid = h.helicsFederateRegisterGlobalEndpoint(cfed, "endpoint1", "")

fid = h.helicsFederateRegisterFilter(cfed, h.helics_filter_type_delay, "filter1")
h.helicsFilterAddSourceTarget(fid, "endpoint1")

h.helicsFilterSet(fid, "delay", 2.0)

# Enter execution mode #
h.helicsFederateEnterExecutingMode(cfed)
print("PI SENDER: Entering execution mode")

# This federate will be publishing deltat*pi for numsteps steps #
this_time = 0.0
value = pi

for t in range(5, 10):
    val = value

    currenttime = h.helicsFederateRequestTime(cfed, t)

    h.helicsPublicationPublishDouble(pub, val)
    print(
        "PI SENDER: Sending value pi = {} at time {} to PI RECEIVER".format(
            val, currenttime
        )
    )
    message = "timestamp=" + str(t)
    h.helicsEndpointSendBytesToAt(epid, message, "endpoint2" , t)
    print(
        "PI SENDER: Sending message '{}' at time {} to PI RECEIVER".format(
            message, currenttime
        )
    )
    time.sleep(1)

h.helicsFederateDisconnect(cfed)
print("PI SENDER: Federate finalized")

while h.helicsBrokerIsConnected(broker):
    time.sleep(1)

h.helicsFederateFree(cfed)
h.helicsCloseLibrary()

print("PI SENDER: Broker disconnected")
